File: Runtime/Plugins/iOS/NuxieUnityBridge.swift
import Foundation
import Combine
#if DEBUG
@_spi(Testing) import Nuxie
#else
import Nuxie
#endif

@objc(NuxieUnityBridge)
public final class NuxieUnityBridge: NSObject {
  @objc public var emit: ((String) -> Void)?
  private var session: String?
  private var snapshotSubscription: AnyCancellable?
  private static weak var owner: NuxieUnityBridge?
  private static var configurationKey: String?
  private lazy var purchases = NuxiePurchaseDelegateBridge { [weak self] name, payload in
    Task { @MainActor [weak self] in self?.send(name, payload) }
  }
  @MainActor private lazy var delegate = NuxieDelegateBridge { [weak self] name, payload in self?.send(name, payload) }

  @objc public func invalidate() {
    Task { @MainActor in
      self.snapshotSubscription?.cancel()
      self.purchases.cancelPending()
      self.session = nil
      self.emit = nil
      if Self.owner === self {
        NuxieSDK.shared.delegate = nil
        Self.owner = nil
      }
    }
  }

  @MainActor private func send(_ name: String, _ payload: [String: Any]) {
    guard let session else { return }
    if let json = try? encode(["session": session, "name": name, "payload": payload]) { emit?(json) }
  }

  @MainActor private func snapshot(_ value: FeatureInfo.Snapshot) -> [String: Any] {
    ["identityGeneration": String(value.identityGeneration), "revision": String(value.revision),
     "state": String(describing: value.state), "all": value.all.mapValues(featureAccessDictionary)]
  }

  @objc public func invoke(_ method: String, arguments: [String: Any],
    resolve: @escaping (Any?) -> Void, reject: @escaping (String, String, NSError?) -> Void) {
    Task { @MainActor in
      do {
        if method != "configure" {
          guard Self.owner === self, let session = self.session, arguments["session"] as? String == session else {
            throw failure("sessionExpired", "Configure Nuxie in this Unity runtime before calling its API")
          }
        }
        let sdk = NuxieSDK.shared
        switch method {
        case "configure":
          let input = try object(arguments["configuration"])
          guard input["contract"] as? Int == 1 else { throw failure("incompatibleBridge", "Rebuild the app with the matching Nuxie native module") }
          let session = try required(input, "session")
          let apiKey = try required(input, "apiKey")
          var keyInput = input; keyInput.removeValue(forKey: "session")
          let key = try encode(keyInput)
          if let owner = Self.owner, owner !== self { throw failure("engineAlreadyAttached", "Another runtime owns Nuxie") }
          if let previous = Self.configurationKey, previous != key { throw failure("alreadyConfigured", "Shutdown before changing Nuxie configuration") }
          if sdk.isSetup && Self.configurationKey == nil { throw failure("alreadyConfigured", "Nuxie was configured outside Unity") }
          Self.owner = self
          self.session = session
          self.purchases.cancelPending()
          self.snapshotSubscription?.cancel()
          let config = NuxieConfiguration(apiKey: apiKey)
          config.environment = input["environment"] as? String == "development" ? .development : .production
          switch input["logLevel"] as? String {
          case "verbose": config.logLevel = .verbose
          case "debug": config.logLevel = .debug
          case "info": config.logLevel = .info
          case "error": config.logLevel = .error
          case "none": config.logLevel = .none
          default: config.logLevel = .warning
          }
          config.localeIdentifier = input["localeIdentifier"] as? String
          config.purchaseHandlingMode = input["purchaseHandlingMode"] as? String == "observer" ? .observer : .full
          config.purchaseDelegate = input["externalBilling"] as? Bool == true ? self.purchases : nil
#if DEBUG
          if let endpoint = ProcessInfo.processInfo.environment["NUXIE_UNITY_API_ENDPOINT"], let url = URL(string: endpoint) {
            config.testingOverrides.apiEndpoint = url
          }
#endif
          sdk.delegate = self.delegate
          if sdk.isSetup { try sdk.setPurchaseDelegate(config.purchaseDelegate) }
          else { try sdk.setup(with: config) }
          Self.configurationKey = key
          self.snapshotSubscription = sdk.features.$snapshot.sink { [weak self] value in
            guard let self else { return }
            self.send("features", self.snapshot(value))
          }
          resolve(try encode(["contract": 1, "session": session, "nativeVersion": sdk.version,
            "snapshot": self.snapshot(sdk.features.snapshot)]))
        case "shutdown":
          self.snapshotSubscription?.cancel()
          self.purchases.cancelPending()
          await sdk.shutdown()
          sdk.delegate = nil
          self.session = nil; Self.owner = nil; Self.configurationKey = nil
          resolve(nil)
        case "identify":
          let options = try object(arguments["properties"])
          sdk.identify(try required(arguments, "customerId"), userProperties: options["properties"] as? [String: Any],
            userPropertiesSetOnce: options["propertiesSetOnce"] as? [String: Any])
          resolve(try encode(self.snapshot(sdk.features.snapshot)))
        case "reset": sdk.reset(keepAnonymousId: false); resolve(try encode(self.snapshot(sdk.features.snapshot)))
        case "getIdentity": resolve(try encode(["distinctId": sdk.getDistinctId(), "anonymousId": sdk.getAnonymousId(), "isIdentified": sdk.isIdentified]))
        case "setLocaleIdentifier": try await sdk.setLocaleIdentifier(arguments["locale"] as? String); resolve(nil)
        case "trigger": sdk.trigger(try required(arguments, "event"), properties: try object(arguments["properties"])); resolve(nil)
        case "dismiss": await sdk.dismiss(); resolve(nil)
        case "hasFeature":
          let options = try object(arguments["options"])
          let access = try await sdk.hasFeature(try required(arguments, "featureId"),
            requiredBalance: options["requiredBalance"] as? Double ?? 1, entityId: options["entityId"] as? String,
            policy: options["policy"] as? String == "remote" ? .remote : .cacheFirst)
          resolve(try encode(featureAccessDictionary(access)))
        case "consumeFeature":
          let options = try object(arguments["options"])
          let result = try await sdk.consumeFeature(try required(arguments, "featureId"),
            quantity: options["quantity"] as? Double ?? 1, operationId: try required(options, "operationId"),
            entityId: options["entityId"] as? String)
          resolve(try encode(["customerId": result.customerId, "featureId": result.featureId, "occurredAtMs": nuxieNullable(result.occurredAtMs),
            "operationId": result.operationId, "accepted": result.accepted, "code": result.code,
            "quantity": result.quantity, "balance": nuxieNullable(result.balance), "unlimited": result.unlimited,
            "active": result.active, "idempotentReplay": result.idempotentReplay]))
        case "restorePurchases":
          let result = await sdk.restorePurchases()
          switch result {
          case .restored: resolve(try encode(["type": "restored"]))
          case .noPurchases: resolve(try encode(["type": "noPurchases"]))
          case .failed: resolve(try encode(["type": "failed", "message": "Restore failed"]))
          }
        case "completePurchase": self.purchases.completePurchase(requestId: try required(arguments, "requestId"), payload: try object(arguments["result"])); resolve(nil)
        case "completeRestore": self.purchases.completeRestore(requestId: try required(arguments, "requestId"), payload: try object(arguments["result"])); resolve(nil)
        default: throw failure("invalidMethod", "Unsupported bridge method")
        }
      } catch {
        if method == "configure", !NuxieSDK.shared.isSetup {
          self.snapshotSubscription?.cancel(); self.purchases.cancelPending()
          self.session = nil
          if Self.owner === self { Self.owner = nil; NuxieSDK.shared.delegate = nil }
        }
        let error = error as NSError
        reject(error.userInfo["code"] as? String ?? "nativeError", error.localizedDescription, error)
      }
    }
  }
}

private func failure(_ code: String, _ message: String) -> NSError {
  NSError(domain: "ai.nuxie.unity", code: 1, userInfo: ["code": code, NSLocalizedDescriptionKey: message])
}
private func required(_ object: [String: Any], _ key: String) throws -> String {
  guard let value = object[key] as? String, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
    throw failure("invalidArgument", "Missing or empty \(key)")
  }
  return value
}
private func object(_ raw: Any?) throws -> [String: Any] {
  guard let raw = raw as? String, let data = raw.data(using: .utf8),
    let value = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
    throw failure("invalidArgument", "Expected a JSON object")
  }
  return value
}
private func encode(_ value: [String: Any]) throws -> String {
  String(decoding: try JSONSerialization.data(withJSONObject: value, options: [.sortedKeys]), as: UTF8.self)
}

File: Native~/Android/src/main/kotlin/ai/nuxie/unity/NuxieUnityBridge.kt
package ai.nuxie.unity

import ai.nuxie.sdk.*
import ai.nuxie.sdk.billing.PurchaseHandlingMode
import ai.nuxie.sdk.billing.RestoreResult
import ai.nuxie.sdk.features.FeatureAccess
import ai.nuxie.sdk.features.FeatureCheckPolicy
import ai.nuxie.sdk.features.FeatureInfo
import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject
import java.lang.ref.WeakReference

class NuxieUnityBridge(activity: android.app.Activity, private val callback: Callback) {
  interface Callback { fun onMessage(message: String) }
  private val activityReference = WeakReference(activity)
  private val context = activity.applicationContext
  companion object {
    const val NAME = "Nuxie"
    private var owner = WeakReference<NuxieUnityBridge>(null)
    private var configurationKey: String? = null
  }
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private var session: String? = null
  private var snapshotJob: Job? = null
  private val purchases = NuxiePurchaseDelegateBridge(emitEvent = { name, value -> scope.launch { send(name, value) } })
  private val listener = object : NuxieListener {
    override fun onActivityEmitted(sdk: Nuxie, info: NuxieActivityInfo) {
      scope.launch { send("activity", mapOf("schemaVersion" to NuxieActivityInfo.SCHEMA_VERSION,
        "id" to info.id, "timestampMs" to info.timestampMillis, "receivedAtMs" to info.receivedAtMillis,
        "name" to info.name, "properties" to info.properties.mapValues { (_, v) -> when(v) {
          is NuxieActivityValue.String -> v.value; is NuxieActivityValue.Int -> v.value
          is NuxieActivityValue.Double -> v.value; is NuxieActivityValue.Bool -> v.value
        } })) }
    }
    override fun onAppActionRequested(sdk: Nuxie, action: AppAction) {
      scope.launch { send("appAction", mapOf("name" to action.name,
        "payload" to action.payload?.mapValues { (_, v) -> when(v) {
          is AppActionValue.String -> v.value; is AppActionValue.Int -> v.value
          is AppActionValue.Double -> v.value; is AppActionValue.Bool -> v.value
        } }, "experience" to mapOf("experienceId" to action.experience.experienceId,
          "experienceVersion" to action.experience.experienceVersion, "journeyId" to action.experience.journeyId))) }
    }
  }
  fun invalidate() {
    scope.launch {
      snapshotJob?.cancel(); purchases.cancelPending("runtime_invalidated")
      session = null
      if (owner.get() === this@NuxieUnityBridge) { owner.clear(); if (Nuxie.listener === listener) Nuxie.listener = null }
      scope.cancel()
    }
  }
  private class Promise(private val id: String, private val callback: Callback) {
    fun resolve(value: Any?) { callback.onMessage(JSONObject(mapOf("requestId" to id, "result" to (value ?: JSONObject.NULL))).toString()) }
    fun reject(code: String, message: String?, error: Exception) { callback.onMessage(JSONObject(mapOf("requestId" to id, "error" to mapOf("code" to code, "message" to (message ?: "Native operation failed")))).toString()) }
  }
  fun dispatch(json: String) {
    val request = JSONObject(json)
    val promise = Promise(request.getString("requestId"), callback)
    try {
      val a = request.getJSONObject("arguments")
      val s = a.optString("session")
      when(request.getString("method")) {
        "configure" -> configure(a.getString("configuration"), promise)
        "shutdown" -> shutdown(s, promise)
        "identify" -> identify(s, a.getString("customerId"), a.getString("properties"), promise)
        "reset" -> reset(s, false, promise)
        "getIdentity" -> getIdentity(s, promise)
        "setLocaleIdentifier" -> setLocaleIdentifier(s, a.nullableString("locale"), promise)
        "trigger" -> trigger(s, a.getString("event"), a.getString("properties"), promise)
        "dismiss" -> dismiss(s, promise)
        "hasFeature" -> hasFeature(s, a.getString("featureId"), a.getString("options"), promise)
        "consumeFeature" -> consumeFeature(s, a.getString("featureId"), a.getString("options"), promise)
        "restorePurchases" -> restorePurchases(s, promise)
        "completePurchase" -> completePurchase(s, a.getString("requestId"), a.getString("result"), promise)
        "completeRestore" -> completeRestore(s, a.getString("requestId"), a.getString("result"), promise)
        else -> throw BridgeFailure("invalidMethod", "Unknown bridge method")
      }
    } catch (e: Exception) { promise.reject("invalidArgument", e.message, e) }
  }
  private fun send(name: String, payload: Map<String, Any?>) {
    val attached = session ?: return
    callback.onMessage(JSONObject(mapOf("session" to attached, "name" to name, "payload" to payload)).toString())
  }
  private fun run(promise: Promise, attached: String? = null, block: suspend () -> Any?) {
    scope.launch {
      try {
        if (attached != null && (attached != session || owner.get() !== this@NuxieUnityBridge)) throw BridgeFailure("sessionExpired", "Configure Nuxie in this runtime first")
        promise.resolve(block())
      } catch (error: Exception) { promise.reject((error as? BridgeFailure)?.code ?: "nativeError", error.message, error) }
    }
  }
  private fun configure(configuration: String, promise: Promise) = run(promise) {
    if (!android.os.Build.SUPPORTED_ABIS.any { it == "arm64-v8a" || it == "x86_64" }) throw BridgeFailure("unsupportedArchitecture", "Nuxie requires a 64-bit Android device")
    val input = JSONObject(configuration)
    if (input.getInt("contract") != 1) throw BridgeFailure("incompatibleBridge", "Rebuild the app with the matching native SDK")
    val attached = input.getString("session")
    val apiKey = input.getString("apiKey")
    require(apiKey.isNotBlank() && attached.isNotBlank())
    input.remove("session")
    val key = JSONObject(input.toMap().toSortedMap()).toString()
    if (owner.get() != null && owner.get() !== this) throw BridgeFailure("engineAlreadyAttached", "Another runtime owns Nuxie")
    if (configurationKey != null && configurationKey != key) throw BridgeFailure("alreadyConfigured", "Shutdown before changing configuration")
    if (Nuxie.isSetup && configurationKey == null) throw BridgeFailure("alreadyConfigured", "Nuxie was configured outside Unity")
    owner = WeakReference(this); session = attached
    purchases.cancelPending("session_replaced"); snapshotJob?.cancel()
    try {
      val activity = activityReference.get()?.takeUnless { it.isDestroyed }
        ?: throw BridgeFailure("missingActivity", "The Unity Activity is no longer available")
      val config = NuxieConfiguration(apiKey = apiKey).apply {
        environment = if (input.optString("environment") == "development") NuxieEnvironment.DEVELOPMENT else NuxieEnvironment.PRODUCTION
        logLevel = when(input.optString("logLevel")) {
          "verbose" -> LogLevel.DEBUG; "debug" -> LogLevel.DEBUG; "info" -> LogLevel.INFO
          "error" -> LogLevel.ERROR; "none" -> LogLevel.NONE; else -> LogLevel.WARN
        }
        localeIdentifier = input.nullableString("localeIdentifier")
        purchaseHandlingMode = if (input.optString("purchaseHandlingMode") == "observer") PurchaseHandlingMode.APP_MANAGED else PurchaseHandlingMode.NUXIE_MANAGED
        purchaseDelegate = if (input.optBoolean("externalBilling")) purchases else null
      }
      if ((context.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) != 0) {
        activity.intent?.getStringExtra("NUXIE_UNITY_API_ENDPOINT")?.let {
          config.testingOverrides.apiEndpoint = java.net.URL(it)
        }
      }
      Nuxie.listener = listener
      if (Nuxie.isSetup) Nuxie.setPurchaseDelegate(config.purchaseDelegate)
      else Nuxie.setup(activity, config)
      configurationKey = key
      snapshotJob = scope.launch { Nuxie.features.snapshot.collect { send("features", it.toWire()) } }
      JSONObject(mapOf("session" to attached, "contract" to 1, "nativeVersion" to Nuxie.version,
        "snapshot" to Nuxie.features.snapshot.value.toWire())).toString()
    } catch (error: Exception) {
      if (!Nuxie.isSetup) { owner.clear(); session = null; if (Nuxie.listener === listener) Nuxie.listener = null }
      throw error
    }
  }
  private fun shutdown(session: String, promise: Promise) = run(promise, session) {
    snapshotJob?.cancel(); purchases.cancelPending("shutdown")
    withContext(Dispatchers.Default) { Nuxie.shutdown() }
    if (Nuxie.listener === listener) Nuxie.listener = null
    this.session = null; owner.clear(); configurationKey = null; null
  }
  private fun identify(session: String, customerId: String, properties: String, promise: Promise) = run(promise, session) {
    val options = JSONObject(properties)
    Nuxie.identify(customerId, options.optJSONObject("properties")?.toMap(), options.optJSONObject("propertiesSetOnce")?.toMap()); JSONObject(Nuxie.features.snapshot.value.toWire()).toString()
  }
  private fun reset(session: String, keepAnonymousId: Boolean, promise: Promise) = run(promise, session) { Nuxie.reset(false); JSONObject(Nuxie.features.snapshot.value.toWire()).toString() }
  private fun getIdentity(session: String, promise: Promise) = run(promise, session) {
    JSONObject(mapOf("distinctId" to Nuxie.distinctId, "anonymousId" to Nuxie.anonymousId, "isIdentified" to Nuxie.isIdentified)).toString()
  }
  private fun setLocaleIdentifier(session: String, locale: String?, promise: Promise) = run(promise, session) { Nuxie.setLocaleIdentifier(locale); null }
  private fun trigger(session: String, event: String, properties: String, promise: Promise) = run(promise, session) { Nuxie.trigger(event, JSONObject(properties).toMap()); null }
  private fun dismiss(session: String, promise: Promise) = run(promise, session) { Nuxie.dismiss(); null }
  private fun hasFeature(session: String, featureId: String, options: String, promise: Promise) = run(promise, session) {
    val input = JSONObject(options)
    val result = Nuxie.hasFeature(featureId, input.optDouble("requiredBalance", 1.0), input.nullableString("entityId"),
      if (input.optString("policy") == "remote") FeatureCheckPolicy.REMOTE else FeatureCheckPolicy.CACHE_FIRST)
    JSONObject(result.toWire()).toString()
  }
  private fun consumeFeature(session: String, featureId: String, options: String, promise: Promise) = run(promise, session) {
    val input = JSONObject(options)
    val result = Nuxie.consumeFeature(featureId, input.getDouble("quantity"), input.getString("operationId"), input.nullableString("entityId"))
    JSONObject(mapOf("customerId" to result.customerId, "featureId" to result.featureId, "occurredAtMs" to result.occurredAtMs,
      "operationId" to result.operationId, "accepted" to result.accepted, "code" to result.code,
      "quantity" to result.quantity, "balance" to result.balance, "unlimited" to result.unlimited,
      "active" to result.active, "idempotentReplay" to result.idempotentReplay)).toString()
  }
  private fun restorePurchases(session: String, promise: Promise) = run(promise, session) {
    JSONObject(when(Nuxie.restorePurchases()) {
      RestoreResult.Restored -> mapOf("type" to "restored")
      RestoreResult.NoPurchases -> mapOf("type" to "noPurchases")
      is RestoreResult.Failed -> mapOf("type" to "failed", "message" to "Restore failed")
    }).toString()
  }
  private fun completePurchase(session: String, requestId: String, result: String, promise: Promise) = run(promise, session) { purchases.completePurchase(requestId, JSONObject(result).toMap()); null }
  private fun completeRestore(session: String, requestId: String, result: String, promise: Promise) = run(promise, session) { purchases.completeRestore(requestId, JSONObject(result).toMap()); null }
}
private class BridgeFailure(val code: String, message: String) : IllegalStateException(message)
private fun FeatureAccess.toWire() = mapOf("allowed" to allowed, "unlimited" to unlimited, "balance" to balance, "type" to when(type) {
  ai.nuxie.sdk.features.FeatureType.BOOLEAN -> "boolean"
  ai.nuxie.sdk.features.FeatureType.METERED -> "metered"
  ai.nuxie.sdk.features.FeatureType.CREDIT_SYSTEM -> "creditSystem"
})
private fun FeatureInfo.Snapshot.toWire(): Map<String, Any?> = mapOf("identityGeneration" to identityGeneration.toString(), "revision" to revision.toString(),
  "state" to when(state) { FeatureInfo.State.Unknown -> "unknown"; FeatureInfo.State.Ready -> "ready"; FeatureInfo.State.Reconciling -> "reconciling" },
  "all" to all.mapValues { it.value.toWire() })
private fun JSONObject.nullableString(key: String): String? = if (isNull(key)) null else getString(key)
private fun JSONObject.toMap(): Map<String, Any?> = keys().asSequence().associateWith { key -> jsonValue(get(key)) }
private fun jsonValue(value: Any?): Any? = when(value) {
  JSONObject.NULL -> null
  is JSONObject -> value.toMap()
  is JSONArray -> (0 until value.length()).map { jsonValue(value.get(it)) }
  else -> value
}
